### ISSUE TITLE
bug(profile) inconsistent profile isLoaded, isEmpty states

### ISSUE THREAD

**Issue by Modestas** [prescottprue/react-redux-firebase#301]
**Do you want to request a *feature* or report a *bug*?**  
Bug

**What is the current behavior?**
In one of the prop states during login process, I get such result:

**no auth:**
 isLoaded(auth) `true` isEmpty(auth) `true` isLoaded(profile) `true` isEmpty(profile) `true`
**after login pressed**
 isLoaded(auth) `false` isEmpty(auth) `true` isLoaded(profile) `false` isEmpty(profile) `true`
**[corrupted]** 
isLoaded(auth) `true` isEmpty(auth) `false` isLoaded(profile) `true` isEmpty(profile) `true`
**on next prop, it is auto fixed:**
isLoaded(auth) `true` isEmpty(auth) `false` isLoaded(profile) `true` isEmpty(profile) `false`

However this corrupted state means that user is authenticated but profile is not found (profile path existed in database even before login).

**What is the expected behavior?**
Corrupted state should not occur. In example, in my app I perform redirect if user is `auth` but does not have an entry in database.

**Which versions of dependencies, and which browser and OS are affected by this issue? Did this work in previous versions or setups?**
2.0.10

**Comment by prescottprue** [comment]
@Modestas Going to investigate further, but from what you are reporting this seems like a bug. Thanks for reporting.

**Comment by prescottprue** [comment]
@Modestas Upon further investigation it seems that this may be expected because you will first be logged in, then your profile listener will be applied, then finally the profile response will come back.

That means that `auth` will always come back before `profile`.

Could you look at which actions are being dispatched in redux devtools so we can confirm?

**Comment by Modestas** [comment]
Action: `{ type: '@@reactReduxFirebase/LOGIN', auth: {...} }`
State: `auth: { isLoaded: true, isEmpty: false }`, `profile: { isLoaded: true, isEmpty: true }`

Action: `{ type: '@@reactReduxFirebase/SET_PROFILE', profile: {...} }`
State: `auth: { isLoaded: true, isEmpty: false }`, `profile: { isLoaded: true, isEmpty: false }`

Maybe this is fine, but in that `LOGIN` moment, state is kind of **invalid,** because it says that `profile` is indeed loaded but empty. In my opinion, login initialization should invalidate `profile`. Yes, it was loaded before login was initialized, but the moment login was started, it should change to `{ isLoaded: false }`.

However, if you want it to work like that, so be it. Maybe there is another way/function to check if profile variable is not prepared yet? Or should I do manual quering to check if user has profile?

**Comment by prescottprue** [comment]
That is a **GREAT** point with having profile set to loaded false by the `LOGIN` action.

Going to change that now, and it should be ready for the next v2.0.0 pre-release.

**Comment by prescottprue** [comment]
[`v2.0.0-beta.12`](https://github.com/prescottprue/react-redux-firebase/releases/tag/v2.0.0-beta.12) was released with profile not being set to `isLoaded: true` unless profile is actually loaded (`LOGIN_ERROR` action no longer affects profile state).

Let me know if it is not functioning as you expect. Thank you for helping test out v2!

**Comment by prescottprue** [comment]
[v2.0.0-beta.13](https://github.com/prescottprue/react-redux-firebase/releases/tag/v2.0.0-beta.13) also includes updates to `profile` state (updating it to loaded on empty Auth change). Not really related to this issue, but may be the solution for some finding themselves here.

**Comment by prescottprue** [comment]
@Modestas Yes, it does still set that way. This is to correctly match functionality of `v1` which set `profile: null` (i.e. loaded) when auth empty changed.

It seems that this is expected by some using this in production (including my company). I am currently working on a config option to disable this (thinking `loadProfileOnEmptyAuthChange`), but it will include notation about the deviation from v1.

Open to ideas here, but it seems like making this the default will need to wait for.

Going to reopen until config option is added. Do you have ideas for names other than `loadProfileOnEmptyAuthChange: false` for disabling?

**Comment by Modestas** [comment]
> This is to correctly match functionality of v1 which set profile: null (i.e. loaded) when auth empty changed.

It makes sence to set it that way because when there is no **auth**, **profile** is non existant.

From my point of view, **profile** is solely and fully dependant from **auth**. Correct me if I'm wrong but it cannot exists without **auth**? So I can't think of a reason why `@@reactReduxFirebase/LOGIN` action should not invalidate profile.

Like this `{ type: '@@reactReduxFirebase/LOGIN', auth: {...} profile: null }`. Do you see some problems with this approach? Is there any use case that fails with this?

**Comment by prescottprue** [comment]
@Modestas I agree with what you are saying (assuming you meant "fully dependent *on* auth" and not "fully *independent* from auth").

I will give this a try and see if that prevents things from breaking on our end.

**Comment by prescottprue** [comment]
@Modestas This will most likely live under another action type (maybe `UNLOAD_PROFILE`), but I was able to confirm that this shouldn't be breaking for us.

When `login` or any other method that could attach/detach the profile listener is called, this new action will be called. That way, profile will be marked as `isLoaded: false` right away.

**Comment by prescottprue** [comment]
As of [the v2.0.0-beta.15 release](https://github.com/prescottprue/react-redux-firebase/releases/tag/v2.0.0-beta.15), the `UNLOAD_PROFILE` action type is dispatched when calling the `login` action creator. This causes the `profile` reducer to switch `state.profile.isLoaded` to `false`. This "invalidates the state on login" as you were mentioning, still investigating doing this with all auth methods.

Let me know if it doesn't function as expected or if you find any other inconsistency (even if just by opinion) within state.

**Comment by Modestas** [comment]
Same result since `AUTH_EMPTY_CHANGE` received after UNLOAD_PROFILE.

`@@INIT`
`@@reactReduxFirebase/AUTHENTICATION_INIT_STARTED`
`@@reactReduxFirebase/AUTHENTICATION_INIT_FINISHED`
`@@reactReduxFirebase/LOGIN_ERROR`
`@@reactReduxFirebase/UNLOAD_PROFILE`
`@@reactReduxFirebase/AUTH_EMPTY_CHANGE`
`@@reactReduxFirebase/LOGIN`

**Comment by prescottprue** [comment]
@Modestas It hasn't been released yet, but [v2.0.0-beta.17](https://github.com/prescottprue/react-redux-firebase/pull/334) is changed so `LOGIN` now sets profile to `isLoaded: false`. It also adds a `preserveOnLogin` config option in case you would like you preserve auth or profile state on login:

```js
preserveOnLogin: {
    profile: ['isLoaded'] // preserve profile.isLoaded state from before LOGIN action
}
```

**Comment by prescottprue** [comment]
[v2.0.0-beta.17](https://github.com/prescottprue/react-redux-firebase/releases/tag/v2.0.0-beta.17) includes a change where the `LOGIN` action changes `profile` state to `isLoaded: false` with a unit test to confirm that this is happening.

If I misunderstood or this solution does not work for you, please let me know. The migration guide has not yet been updated with this yet, but it will most likely need to be.
